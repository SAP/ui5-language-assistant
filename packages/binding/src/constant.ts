export const BINDING_ISSUE_TYPE = "binding-issue";
export const PROPERTY_BINDING_INFO =
  "sap.ui.base.ManagedObject.PropertyBindingInfo";
export const AGGREGATION_BINDING_INFO =
  "sap.ui.base.ManagedObject.AggregationBindingInfo";
export const Binding_Mode = "sap.ui.model.BindingMode";
