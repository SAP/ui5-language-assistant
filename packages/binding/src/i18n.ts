import type { i18n, TFunction } from "i18next";
import { DEFAULT_I18N_NAMESPACE } from "@ui5-language-assistant/context";
import i18nEn from "./i18n/i18n.json";
import i18next from "i18next";

export function initI18n(i18nInstance: i18n): void {
  i18nInstance.addResourceBundle(
    i18nInstance.language,
    DEFAULT_I18N_NAMESPACE,
    i18nEn
  );
}

export const adaptTranslatedText = (text: string): string => {
  if (/&quot;/.test(text)) {
    text = text.replace(/&quot;/g, '"');
  }
  if (/&#39;/.test(text)) {
    text = text.replace(/&#39;/g, "'");
  }
  if (/&amp;quot;/.test(text)) {
    text = text.replace(/&amp;quot;/g, "&quot;");
  }
  return text;
};

export const t: TFunction = (key: string, ...args) => {
  const result = i18next.t(key, ...args);
  return adaptTranslatedText(result);
};
