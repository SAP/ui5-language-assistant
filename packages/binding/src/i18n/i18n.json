{
  "UNKNOWN_CHARACTER": "Unknown character",
  "EXPECT_CLOSING_BRACE": "Expect closing brace",
  "EXPECT_CLOSING_BRACKET": "Expect closing bracket",
  "TOO_MANY_COLON": "Too many colon",
  "EXPECT_COLON": "Expect colon",
  "MISSING_COMMA": "Missing comma",
  "TOO_MANY_COMMAS": "Too many commas",
  "TRAILING_COMMA": "Trailing comma",
  "DUPLICATE_PROPERTY": "Duplicate property",
  "EXPECT_KEY": "Expect key",
  "ALLOWED_VALUES_ARE": "Allowed values are {{data}}",
  "ALLOWED_VALUES_ARE_QUOTES": "Allowed values are \"{{data}}\"",
  "ALLOWED_VALUES_IS": "Allowed value is {{data}}",
  "OR": " or ",
  "COMMA": ", ",
  "REQUIRED_VALUES": "Required values {{data}} must be provided",
  "REQUIRED_VALUE": "Required value {{data}} must be provided",
  "VALID_BINDING_PROPERTY": "A valid binding property info must be provided for \"{}\"",
  "NESTED_COLLECTION": "Nested \"[]\" are not allowed",
  "REQUIRED_DEPENDENCY": "Required dependency \"{{name}}\" should be defined",
  "UNKNOWN_AGGREGATION_BINDING": "Unknown aggregation binding",
  "UNKNOWN_PROPERTY_BINDING": "Unknown property binding info",
  "EXPECT_A_VALUE": "Expect a value",
  "EXPECT_DATA_AS_A_VALUE": "Expect {{data}} as a value",
  "RECURSIVE_COMPOSITE_BINDING": "Recursive composite bindings is not allowed",
  "ONE_OF_THESE_ELEMENTS": "One of these elements [{{data}}] are allowed",
  "MANDATORY_PROPERTY": "Mandatory property \"{{name}}\" must be defined",
  "TYPE": "Type:",
  "DESCRIPTION": "Description:",
  "OPTIONAL": "Optional:",
  "MORE_INFO": "[More information]"
}
