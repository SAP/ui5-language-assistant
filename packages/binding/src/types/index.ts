export type { BindingIssue } from "./issues";
export type {
  PropertyBindingInfoElement,
  Dependents,
  PropertyType,
  BindContext,
} from "./binding-info";
export { BindingInfoName, TypeKind } from "./binding-info";

export { BINDING_ISSUE_TYPE } from "../../src/constant";

export type {
  CursorContext,
  KeyValueContext,
  BaseContext,
  EmptyContext,
  InitialContext,
  KeyContext,
  ValueContext,
} from "./cursor";
