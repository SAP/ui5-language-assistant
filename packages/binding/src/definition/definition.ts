import {
  UI5Class,
  UI5Type,
} from "@ui5-language-assistant/semantic-model-types";
import { PROPERTY_BINDING_INFO } from "../constant";
import {
  BindContext,
  PropertyBindingInfoElement,
  BindingInfoName,
  PropertyType,
  TypeKind,
  Dependents,
} from "../types";
import { ui5NodeToFQN } from "@ui5-language-assistant/logic-utils";
import { forOwn } from "lodash";
import { getDocumentation } from "../utils";
import { fallBackElements } from "./fall-back-definition";

const isBindingInfoName = (name: string): name is BindingInfoName => {
  return !!BindingInfoName[name];
};
const notAllowedElements: Map<BindingInfoName, BindingInfoName[]> = new Map([
  [BindingInfoName.path, [BindingInfoName.parts, BindingInfoName.value]],
  [BindingInfoName.value, [BindingInfoName.parts, BindingInfoName.path]],
  [BindingInfoName.parts, [BindingInfoName.value, BindingInfoName.path]],
]);
const dependents: Map<BindingInfoName, Dependents[]> = new Map([
  [
    BindingInfoName.constraints,
    [
      {
        name: BindingInfoName.type,
        type: [
          {
            kind: TypeKind.string,
            dependents: [],
            notAllowedElements: [],
          },
        ],
      },
    ],
  ],
  [
    BindingInfoName.formatOptions,
    [
      {
        name: BindingInfoName.type,
        type: [
          {
            kind: TypeKind.string,
            dependents: [],
            notAllowedElements: [],
          },
        ],
      },
    ],
  ],
]);
const defaultBoolean: Map<string, boolean[]> = new Map([
  ["Boolean", [true, false]],
  ["boolean", [true, false]],
]);
const getPossibleValuesForClass = (
  context: BindContext,
  type: UI5Class
): string[] => {
  const result: string[] = [];
  forOwn(context.ui5Model.classes, (value, key) => {
    let clzExtends = value.extends;
    while (clzExtends !== undefined) {
      if (clzExtends.name === type.name && clzExtends.kind === type.kind) {
        result.push(key);
        break;
      }
      clzExtends = clzExtends.extends;
    }
  });

  return result;
};

const getFromMap = <T, U extends string>(map: Map<U, T[]>, name: U): T[] => {
  return map.get(name) ?? [];
};

const buildType = (
  context: BindContext,
  type: UI5Type,
  name: BindingInfoName,
  collection = false
): PropertyType[] => {
  const propertyType: PropertyType[] = [];
  switch (type.kind) {
    case "PrimitiveType":
      propertyType.push({
        kind: TypeKind[type.name],
        dependents: getFromMap(dependents, name),
        notAllowedElements: getFromMap(notAllowedElements, name),
        possibleValue: {
          fixed: !!defaultBoolean.get(type.name),
          values: getFromMap(defaultBoolean, type.name),
        },
        collection,
      });
      break;
    case "UI5Enum":
      propertyType.push({
        kind: TypeKind.string,
        dependents: getFromMap(dependents, name),
        notAllowedElements: getFromMap(notAllowedElements, name),
        possibleValue: {
          fixed: true,
          values: type.fields.map((field) => ui5NodeToFQN(field)),
        },
        collection,
      });
      break;
    case "UI5Class":
      propertyType.push({
        kind: TypeKind.string,
        dependents: getFromMap(dependents, name),
        notAllowedElements: getFromMap(notAllowedElements, name),
        possibleValue: {
          fixed: false,
          values: getPossibleValuesForClass(context, type),
        },
        collection,
      });
      break;
    case "UI5Typedef":
      if (TypeKind[type.name]) {
        propertyType.push({
          kind: TypeKind[type.name],
          dependents: getFromMap(dependents, name),
          notAllowedElements: getFromMap(notAllowedElements, name),
          collection,
        });
      }
      break;
    case "UnionType":
      for (const unionType of type.types) {
        propertyType.push(
          ...buildType(context, unionType, name, type.collection)
        );
      }
      break;
    case "ArrayType":
      if (type.type?.kind === "UI5Typedef") {
        propertyType.push(...buildType(context, type.type, name, true));
      }
      break;
  }
  return propertyType;
};

export const getPropertyBindingInfoElements = (
  context: BindContext
): PropertyBindingInfoElement[] => {
  const elements: PropertyBindingInfoElement[] = [];
  const propBinding = context.ui5Model.typedefs[PROPERTY_BINDING_INFO];
  if (!propBinding) {
    return fallBackElements;
  }
  /* istanbul ignore next */
  const properties = propBinding.properties ?? [];
  for (const property of properties) {
    const { name, type } = property;
    if (!isBindingInfoName(name) || !type) {
      /* istanbul ignore next */
      continue;
    }
    const builtType = buildType(context, type, name).reduce(
      (previous: PropertyType[], current: PropertyType) => {
        const index = previous.findIndex((i) => i.kind === current.kind);
        if (index !== -1) {
          // there is duplicate
          /* istanbul ignore next */
          if (current.possibleValue?.values.length !== 0) {
            // has possible value, remove previous - keep current
            return [...previous.slice(index), current];
          }
          /* istanbul ignore next */
          if (previous[index].possibleValue?.values.length !== 0) {
            // has possible value - keep it
            return previous;
          }
        }
        return [...previous, current];
      },
      []
    );
    elements.push({
      name: name,
      type: builtType,
      documentation: getDocumentation(context, property),
    });
  }
  return elements;
};
