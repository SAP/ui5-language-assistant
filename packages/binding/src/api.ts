import type { BindingIssue } from "./types";
import { BINDING_ISSUE_TYPE } from "./types";

export { getCompletionItems } from "./services/completion";
export { bindingValidators } from "./services/diagnostics";
export { getHover } from "./services/hover";
export type { BindingIssue } from "./types";

export function isBindingIssue<T extends { issueType: string }>(
  issue: BindingIssue | T
): issue is BindingIssue {
  return issue.issueType === BINDING_ISSUE_TYPE;
}

export {
  AGGREGATION_BINDING_INFO,
  PROPERTY_BINDING_INFO,
  FILTER_OPERATOR,
} from "./constant";

export { getBindingElements } from "./definition/definition";
