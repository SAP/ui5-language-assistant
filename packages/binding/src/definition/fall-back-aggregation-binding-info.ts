import { BindingInfoName, TypeKind } from "../types";
import type { BindingInfoElement } from "../types";
import { filters } from "./filter";
import { sorter } from "./sorter";

export const aggregationBindingInfoElements: BindingInfoElement[] = [
  {
    name: BindingInfoName.events,
    type: [
      {
        kind: TypeKind.object,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.model,
    type: [
      {
        kind: TypeKind.string,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.path,
    required: true,
    type: [
      {
        kind: TypeKind.string,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.parameters,
    type: [
      {
        kind: TypeKind.object,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.suspended,
    type: [
      {
        kind: TypeKind.boolean,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.factory,
    type: [
      {
        kind: TypeKind.string,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.groupHeaderFactory,
    type: [
      {
        kind: TypeKind.string,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.key,
    type: [
      {
        kind: TypeKind.string,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.startIndex,
    type: [
      {
        kind: TypeKind.integer,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.length,
    type: [
      {
        kind: TypeKind.integer,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.template,
    type: [
      {
        kind: TypeKind.string,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  {
    name: BindingInfoName.templateShareable,
    type: [
      {
        kind: TypeKind.boolean,
        dependents: [],
        notAllowedElements: [],
      },
    ],
    documentation: {
      value: "",
      kind: "plaintext",
    },
  },
  ...filters,
  ...sorter,
];
