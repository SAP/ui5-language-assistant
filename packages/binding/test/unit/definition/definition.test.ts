import { join } from "path";
import {
  Config,
  ProjectName,
  ProjectType,
  TestFramework,
} from "@ui5-language-assistant/test-framework";
import { getContext } from "@ui5-language-assistant/context";
import { BindContext } from "../../../src/types";
import { getBindingElements } from "../../../src/definition/definition";
import { UI5Typedef } from "@ui5-language-assistant/semantic-model-types";
import { initI18n } from "../../../src/i18n";

describe("definition", () => {
  let framework: TestFramework;
  let context: BindContext;
  const viewFilePathSegments = [
    "app",
    "manage_travels",
    "webapp",
    "ext",
    "main",
    "Main.view.xml",
  ];
  beforeAll(async () => {
    const config: Config = {
      projectInfo: {
        name: ProjectName.cap,
        type: ProjectType.CAP,
        npmInstall: true,
        deleteBeforeCopy: false,
      },
    };
    framework = new TestFramework(config);
    const i18n = await framework.initI18n();
    initI18n(i18n);
    context = (await getContext(
      join(framework.getProjectRoot(), ...viewFilePathSegments)
    )) as BindContext;
  });
  describe("getBindingElements", () => {
    it("get binding elements", () => {
      const result = getBindingElements(context);
      expect(result).toMatchSnapshot();
    });
    it("get binding elements - aggregation", () => {
      const result = getBindingElements(context, true);
      expect(result).toMatchSnapshot();
    });
    it("check fallback", () => {
      const result = getBindingElements(
        {
          ...context,
          ui5Model: {
            ...context.ui5Model,
            typedefs: {
              ...context.ui5Model.typedefs,
              "sap.ui.base.ManagedObject.PropertyBindingInfo":
                undefined as unknown as UI5Typedef,
            },
          },
        },
        false
      );
      expect(result).toMatchSnapshot();
    });
    it("check fallback - aggregation", () => {
      const result = getBindingElements(
        {
          ...context,
          ui5Model: {
            ...context.ui5Model,
            typedefs: {
              ...context.ui5Model.typedefs,
              "sap.ui.base.ManagedObject.AggregationBindingInfo":
                undefined as unknown as UI5Typedef,
            },
          },
        },
        true
      );
      expect(result).toMatchSnapshot();
    });
  });
});
