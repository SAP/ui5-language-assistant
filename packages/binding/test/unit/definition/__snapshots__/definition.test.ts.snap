// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`definition getPropertyBindingInfoElements get binding elements 1`] = `
Array [
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** String

**Description:** Path in the model to bind to, either an absolute path or relative to the binding context for the corresponding model; when the path contains a '&gt;' sign, the string preceding it will override the <code>model</code> property and the remainder after the '&gt;' will be used as binding path

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "path",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [
          "parts",
          "value",
        ],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** String

**Description:** Since 1.61, defines a static binding with the given value.

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "value",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [
          "parts",
          "path",
        ],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** String

**Description:** Name of the model to bind against; when <code>undefined</code> or omitted, the default model is used

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "model",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** Boolean

**Description:** Whether the binding should be suspended initially

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "suspended",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "boolean",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": true,
          "values": Array [
            true,
            false,
          ],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** Function

**Description:** Function to convert model data into a property value

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "formatter",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** Boolean

**Description:** Whether the parameters to the formatter function should be passed as raw values. In this case the specified types for the binding parts are not used and the values are not formatted.

<b>Note</b>: use this flag only when using multiple bindings. If you use only one binding and want raw values then simply don't specify a type for that binding.

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "useRawValues",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "boolean",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": true,
          "values": Array [
            true,
            false,
          ],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** Boolean

**Description:** Whether the parameters to the formatter function should be passed as the related JavaScript primitive values. In this case the values of the model are parsed by the {@link sap.ui.model.SimpleType#getModelFormat model format} of the specified types from the binding parts.

<b>Note</b>: use this flag only when using multiple bindings.

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "useInternalValues",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "boolean",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": true,
          "values": Array [
            true,
            false,
          ],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** sap.ui.model.Type | String

**Description:** A type object or the name of a type class to create such a type object; the type will be used for converting model data to a property value (aka \\"formatting\\") and vice versa (in binding mode <code>TwoWay</code>, aka \\"parsing\\")

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "type",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [
            "sap.ca.ui.model.type.Date",
            "sap.ca.ui.model.type.DateTime",
            "sap.ca.ui.model.type.FileSize",
            "sap.ca.ui.model.type.Number",
            "sap.ca.ui.model.type.Time",
            "sap.m.DynamicDate",
            "sap.ui.model.CompositeType",
            "sap.ui.model.odata.type.Boolean",
            "sap.ui.model.odata.type.Byte",
            "sap.ui.model.odata.type.Currency",
            "sap.ui.model.odata.type.Date",
            "sap.ui.model.odata.type.DateTime",
            "sap.ui.model.odata.type.DateTimeBase",
            "sap.ui.model.odata.type.DateTimeOffset",
            "sap.ui.model.odata.type.DateTimeWithTimezone",
            "sap.ui.model.odata.type.Decimal",
            "sap.ui.model.odata.type.Double",
            "sap.ui.model.odata.type.Guid",
            "sap.ui.model.odata.type.Int",
            "sap.ui.model.odata.type.Int16",
            "sap.ui.model.odata.type.Int32",
            "sap.ui.model.odata.type.Int64",
            "sap.ui.model.odata.type.ODataType",
            "sap.ui.model.odata.type.Raw",
            "sap.ui.model.odata.type.SByte",
            "sap.ui.model.odata.type.Single",
            "sap.ui.model.odata.type.Stream",
            "sap.ui.model.odata.type.String",
            "sap.ui.model.odata.type.Time",
            "sap.ui.model.odata.type.TimeOfDay",
            "sap.ui.model.odata.type.Unit",
            "sap.ui.model.SimpleType",
            "sap.ui.model.type.Boolean",
            "sap.ui.model.type.Currency",
            "sap.ui.model.type.Date",
            "sap.ui.model.type.DateInterval",
            "sap.ui.model.type.DateTime",
            "sap.ui.model.type.DateTimeInterval",
            "sap.ui.model.type.FileSize",
            "sap.ui.model.type.Float",
            "sap.ui.model.type.Integer",
            "sap.ui.model.type.String",
            "sap.ui.model.type.Time",
            "sap.ui.model.type.TimeInterval",
            "sap.ui.model.type.Unit",
            "sap.ui.mdc.field.ConditionsType",
            "sap.ui.mdc.field.ConditionType",
            "sap.ui.mdc.field.DynamicDateRangeConditionsType",
          ],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** String

**Description:** Target type to be used by the type when formatting model data, for example \\"boolean\\" or \\"string\\" or \\"any\\"; defaults to the property's type

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "targetType",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** Object

**Description:** Format options to be used for the type; only taken into account when the type is specified by its name - a given type object won't be modified

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "formatOptions",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [
          Object {
            "name": "type",
            "type": Array [
              Object {
                "dependents": Array [],
                "kind": "string",
                "notAllowedElements": Array [],
              },
            ],
          },
        ],
        "kind": "object",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** Object

**Description:** Additional constraints to be used when constructing a type object from a type name, ignored when a type object is given

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "constraints",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [
          Object {
            "name": "type",
            "type": Array [
              Object {
                "dependents": Array [],
                "kind": "string",
                "notAllowedElements": Array [],
              },
            ],
          },
        ],
        "kind": "object",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** sap.ui.model.BindingMode

**Description:** Binding mode to be used for this property binding (e.g. one way)

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "mode",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": true,
          "values": Array [
            "sap.ui.model.BindingMode.Default",
            "sap.ui.model.BindingMode.OneTime",
            "sap.ui.model.BindingMode.OneWay",
            "sap.ui.model.BindingMode.TwoWay",
          ],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** Object

**Description:** Map of additional parameters for this binding; the names and value ranges of the supported parameters depend on the model implementation, they should be documented with the <code>bindProperty</code> method of the corresponding model class or with the model specific subclass of <code>sap.ui.model.PropertyBinding</code>

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "parameters",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "object",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** Object

**Description:** Map of event handler functions keyed by the name of the binding events that they should be attached to

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "events",
    "type": Array [
      Object {
        "collection": false,
        "dependents": Array [],
        "kind": "object",
        "notAllowedElements": Array [],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
    ],
  },
  Object {
    "documentation": Object {
      "kind": "markdown",
      "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`



**Type:** Array<(String | sap.ui.base.ManagedObject.PropertyBindingInfo>)

**Description:** Array of binding info objects for the parts of a composite binding; the structure of each binding info is the same as described for the <code>oBindingInfo</code> as a whole.

If a part is not specified as a binding info object but as a simple string, a binding info object will be created with that string as <code>path</code>. The string may start with a model name prefix (see property <code>path</code>).

<b>Note</b>: recursive composite bindings are currently not supported. Therefore, a part must not contain a <code>parts</code> property.

**Optional:** true

[More information](https://ui5.sap.com/1.108.1/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
    },
    "name": "parts",
    "type": Array [
      Object {
        "collection": true,
        "dependents": Array [],
        "kind": "string",
        "notAllowedElements": Array [
          "value",
          "path",
        ],
        "possibleValue": Object {
          "fixed": false,
          "values": Array [],
        },
      },
      Object {
        "collection": true,
        "dependents": Array [],
        "kind": "object",
        "notAllowedElements": Array [
          "value",
          "path",
        ],
      },
    ],
  },
]
`;
